src/tangents.jl

```diff
@@ -474,45 +448,36 @@ end
     stackdict[x] = zt
     return _map_if_assigned!(Base.Fix2(zero_tangent_internal, stackdict), zt, x)::Array{tangent_type(P), N}
 end
-@inline function zero_tangent_internal(x::P, stackdict::IdDict) where {P<:Union{Tuple, NamedTuple}}
-    return tangent_type(P) == NoTangent ? NoTangent() : tuple_map(Base.Fix2(zero_tangent_internal, stackdict), x)
-end
-function zero_tangent_internal(x::P, stackdict::IdDict) where {P}
-
+function zero_tangent_internal(x::P, stackdict) where {P}
     tangent_type(P) == NoTangent && return NoTangent()
 
     if tangent_type(P) <: MutableTangent
+        stackdict isa IdDict || error("...")
         if haskey(stackdict, x)
             return stackdict[x]::tangent_type(P)
         end
         stackdict[x] = tangent_type(P)() # create a uninitialised MutableTangent
         # if circular reference exists, then the recursive call will first look up the stackdict
         # and return the uninitialised MutableTangent
         # after the recursive call returns, the stackdict will be initialised
-        stackdict[x].fields = backing_type(P)(zero_tangent_struct_field(x, stackdict))
+        stackdict[x].fields = zero_tangent_struct_field(x, stackdict)
         return stackdict[x]::tangent_type(P)
     else
-        if isbitstype(P)
-            return zero_tangent_internal(x)
-        else
-            return tangent_type(P)(backing_type(P)(zero_tangent_struct_field(x, stackdict)))
-        end
+        return tangent_type(P)(zero_tangent_struct_field(x, stackdict))
     end
 end
 
-@inline function zero_tangent_struct_field(x::P, stackdict::IdDict) where {P}
-    return ntuple(fieldcount(P)) do n
+@generated function zero_tangent_struct_field(x::P, stackdict::Union{IdDict, Nothing}) where {P}
```

Comment:
You've gone for `Union{IdDict, Nothing}` here as the restriction on the type of `stackdict`, whereas in other places `Any` has been used. Can we use `Any` here as well, or is there a good reason not to that I'm missing?